modules/javafx.graphics/src/main/java/com/sun/glass/ui/View.java

```diff
@@ -528,17 +528,26 @@ public void setEventHandler(EventHandler eventHandler) {
         this.eventHandler = eventHandler;
     }
 
+    private boolean shouldHandleEvent() {
+        // Don't send any more events if the application has shutdown
+        if (Application.GetApplication() == null) {
```

Comment:
Should we declare `Application.application` field as volatile?

I see interleaved multi-threaded access pattern:

```
thread=JavaFX-Launcher
CREATE thread=JavaFX-Launcher
thread=JavaFX-Launcher
thread=Thread-2
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX-Launcher
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX-Launcher
thread=JavaFX Application Thread
thread=PulseTimer-CVDisplayLink thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=QuantumRenderer-0
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=PulseTimer-CVDisplayLink thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX-Launcher
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX-Launcher
thread=JavaFX-Launcher
thread=JavaFX-Launcher
thread=QuantumRenderer-0
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
thread=JavaFX Application Thread
FINISH thread=JavaFX Application Thread
```

It might be ok though, since both creation and setting to null is done by the FX App thread.

Reply:
Other threads may still not see the current value of the `Application.application` field. Is it safe for other threads to use the `Application` instance after the field has been set to `null`?

Reply:
This is unrelated to the current bug, since all accesses are happening on the FX app thread, so I'd like to file a follow-up to look into this.

To answer the question, there might be a potential problem with other threads not seeing that the `Application.application` field has been set to null.

Other threads do not access the glass Application directly. Most methods requires the app to be on the application thread, and will correctly detect and fail whether they see the old or new value of `Application.application`, so those operations will be fine.

One scenario that should be examined is `Platform.runLater`, which could call into glass and see the old value of application. There are related fields in `Tooklit` (e.g., the `fxUserThread` variable) that ought to be examined at the same time.

Reply:
> since all accesses are happening on the FX app thread

Not quite right: the following threads access the field, apart from JavaFX Application Thread:

```
thread=JavaFX-Launcher
thread=Thread-2
thread=PulseTimer-CVDisplayLink thread
thread=QuantumRenderer-0
```

The output above logs thread names that access `Application.application` field, running the reproducer attached to both tickets.

Reply:
To clarify, I'm talking about the accesses that have anything to do with this bug, and specifically, the methods that I modified. Those are all on the FX application thread. Unrelated to this bug, there are plenty of calls to Platform.runLater on other threads, which will indeed access the `Application.application` field from that other thread.

Reply:
I filed [JDK-8336476](https://bugs.openjdk.org/browse/JDK-8336476) as a follow-up bug to address the race condition.